File: frontend/src/services/activiteService.ts
import { get, post, put, patch, del } from "./api";
import type { ActiviteEconomique, PaginatedResponse, PaginationParams } from "../types";

function qs(params: PaginationParams): Record<string, string | number | boolean | undefined> {
  return {
    page: params.page,
    per_page: params.per_page,
    search: params.search,
    sort: params.sort,
    order: params.order,
  };
}

export const activiteService = {
  list(params: PaginationParams = {}): Promise<PaginatedResponse<ActiviteEconomique>> {
    return get<PaginatedResponse<ActiviteEconomique>>("/activites-economiques", qs(params));
  },

  get(id: number): Promise<ActiviteEconomique> {
    return get<ActiviteEconomique>(`/activites-economiques/${id}`);
  },

  create(data: Partial<ActiviteEconomique>): Promise<ActiviteEconomique> {
    return post<ActiviteEconomique>("/activites-economiques", data);
  },

  update(id: number, data: Partial<ActiviteEconomique>): Promise<ActiviteEconomique> {
    return put<ActiviteEconomique>(`/activites-economiques/${id}`, data);
  },

  delete(id: number): Promise<void> {
    return del<void>(`/activites-economiques/${id}`);
  },

  secteurs(): Promise<string[]> {
    return get<string[]>("/activites-economiques/secteurs");
  },

  statistiques(): Promise<Record<string, unknown>> {
    return get<Record<string, unknown>>("/activites-economiques/statistiques");
  },
};

File: gs_operateur_mobile/lib/services/activite_service.dart
import '../models/activite.dart';
import 'api_service.dart';

class ActiviteService {
  final ApiService _api = ApiService();

  Future<List<ActiviteEconomique>> list({String? search, int? personneId, int page = 1}) async {
    final response = await _api.get('/activites-economiques', queryParameters: {
      if (search != null && search.isNotEmpty) 'search': search,
      if (personneId != null) 'personne_id': personneId,
      'page': page,
    });

    if (response['success'] == true) {
      final data = response['data'];
      return (data is List ? data : data['data'] ?? [])
          .map<ActiviteEconomique>((a) => ActiviteEconomique.fromJson(a))
          .toList();
    }
    throw Exception(response['message'] ?? 'Erreur');
  }

  Future<ActiviteEconomique> get(int id) async {
    final response = await _api.get('/activites-economiques/$id');
    if (response['success'] == true) {
      return ActiviteEconomique.fromJson(response['data']);
    }
    throw Exception(response['message'] ?? 'Erreur');
  }

  Future<ActiviteEconomique> create(Map<String, dynamic> data) async {
    final response = await _api.post('/activites-economiques', data: data);
    if (response['success'] == true) {
      return ActiviteEconomique.fromJson(response['data']);
    }
    throw Exception(response['message'] ?? 'Erreur lors de la création');
  }

  Future<ActiviteEconomique> update(int id, Map<String, dynamic> data) async {
    final response = await _api.put('/activites-economiques/$id', data: data);
    if (response['success'] == true) {
      return ActiviteEconomique.fromJson(response['data']);
    }
    throw Exception(response['message'] ?? 'Erreur lors de la modification');
  }

  Future<void> delete(int id) async {
    final response = await _api.delete('/activites-economiques/$id');
    if (response['success'] != true) {
      throw Exception(response['message'] ?? 'Erreur lors de la suppression');
    }
  }
}
